src/MQTTClient.h

```diff
@@ -1,63 +1,323 @@
 #ifndef MQTT_CLIENT_H
 #define MQTT_CLIENT_H
 
-#ifndef MQTT_BUFFER_SIZE
-#define MQTT_BUFFER_SIZE 128
-#endif
-
-#define MQTTCLIENT_QOS1 0
-#define MQTTCLIENT_QOS2 0
-
-#include "Network.h"
-#include "Timer.h"
-#include "lib/MQTTClient.h"
 #include <Arduino.h>
 #include <Client.h>
 #include <Stream.h>
 
+#include "system.h"
+
+typedef void (*MQTTClientCallbackSimple)(String topic, String payload);
+typedef void (*MQTTClientCallbackAdvanced)(String topic, String payload, char bytes[], unsigned int length);
+
 typedef struct {
-  char *topic;
-  char *payload;
-  unsigned int length;
-  boolean retained;
-} MQTTMessage;
-
-void messageReceived(String topic, String payload, char *bytes,
-                     unsigned int length);
-
-class MQTTClient {
-private:
-  Network network;
-  MQTT::Client<Network, Timer, MQTT_BUFFER_SIZE, 0> *client;
-  MQTTPacket_connectData options;
+    bool use_advanced = false;
+    MQTTClientCallbackSimple simple = NULL;
+    MQTTClientCallbackAdvanced advanced = NULL;
+} MQTTClientCallback;
+
+static void MQTTClient_callback(lwmqtt_client_t *client, void *ref, lwmqtt_string_t *topic, lwmqtt_message_t *message) {
+  // get callback
+  MQTTClientCallback * cb = (MQTTClientCallback*)ref;
+
+  // null terminate topic to create String object
+  char t[topic->len + 1];
+  memcpy(t, topic->data, (size_t)topic->len);
+  t[topic->len] = '\0';
+
+  // get payload
+  char *payload = (char *)message->payload;
+
+  // null terminate payload
+  payload[message->payload_len] = '\0';
+
+  // call the user callback
+  if(cb->use_advanced) {
+    cb->advanced(String(t), String(payload), (char *)message->payload, (unsigned int)message->payload_len);
+  } else {
+    cb->simple(String(t), String(payload));
+  }
+}
+
+template <int BUF_SIZE>
```

Comment:
Hmm, any reason to use templates here instead of dynamic memory allocation (via `malloc`)? You generally don't see templates used in Arduino libraries and sketches.